code/modules/research/experimentor.dm

```diff
@@ -142,117 +134,115 @@
-		updateUsrDialog()
-	else
-		if(recentlyExperimented)
-			to_chat(usr, span_warning("[src] has been used too recently!"))
-		else if(!loaded_item)
-			to_chat(usr, span_warning("[src] is not currently loaded!"))
-		else if(!process || process != loaded_item) //Interface exploit protection (such as hrefs or swapping items with interface set to old item)
-			to_chat(usr, span_danger("Interface failure detected in [src]. Please try again."))
-		else
-			var/dotype
-			if(text2num(scantype) == SCANTYPE_DISCOVER)
-				dotype = SCANTYPE_DISCOVER
-			else
-				dotype = matchReaction(process,scantype)
-			experiment(dotype,process)
-			use_power(750)
-			if(dotype != FAIL)
-				var/list/nodes = techweb_item_unlock_check(process)
-				var/picked = pick_weight(nodes) //This should work.
-				stored_research.unhide_node(SSresearch.techweb_node_by_id(picked))
-	updateUsrDialog()
-
-/obj/machinery/rnd/experimentor/proc/matchReaction(matching,reaction)
-	var/obj/item/D = matching
-	if(D)
-		var/list/item_reactions = item_reactions()
-		if(item_reactions.Find("[D.type]"))
-			var/tor = item_reactions["[D.type]"]
-			if(tor == text2num(reaction))
-				return tor
-			else
-				return FAIL
-		else
-			return FAIL
-	else
+
+	if(delete)
+		QDEL_NULL(loaded_item)
+		return
+
+	var/atom/drop_atom = get_step(src, EAST) || drop_location()
+	if(cloneMode)
+		visible_message(span_notice("A duplicate of \the [loaded_item] pops out!"))
+		new loaded_item.type(drop_atom)
+		cloneMode = FALSE
+		return
+
+	loaded_item.forceMove(drop_atom)
+	loaded_item = null
+
+/obj/machinery/rnd/experimentor/proc/_match_reaction(obj/item/matching, target_reaction)
+	if(isnull(matching) | isnull(target_reaction))
 		return FAIL
 
-/obj/machinery/rnd/experimentor/proc/ejectItem(delete=FALSE)
-	if(loaded_item)
-		if(cloneMode)
-			visible_message(span_notice("A duplicate [loaded_item] pops out!"))
-			var/type_to_make = loaded_item.type
-			new type_to_make(get_turf(pick(oview(1,src))))
-			cloneMode = FALSE
-			return
-		var/turf/dropturf = get_turf(pick(view(1,src)))
-		if(!dropturf) //Failsafe to prevent the object being lost in the void forever.
-			dropturf = drop_location()
-		loaded_item.forceMove(dropturf)
-		if(delete)
-			qdel(loaded_item)
-		loaded_item = null
+	var/list/item_reactions = item_reactions()
+	if("[matching.type]" in item_reactions)
+		var/associated_reaction = item_reactions["[matching.type]"]
+		if(associated_reaction == target_reaction)
+			return associated_reaction
+
+	return FAIL
+
+/obj/machinery/rnd/experimentor/proc/_try_perform_experiment(reaction_id)
+	if(isnull(stored_research))
+		balloon_alert_to_viewers("not connected to server!")
+		return
+
+	if(recentlyExperimented)
+		balloon_alert_to_viewers("used too recently!")
+		return
+
+	if(isnull(loaded_item))
+		return
+
+	var/reaction = text2num(reaction_id)
+	if(isnull(reaction))
+		return
+
+	if(reaction != SCANTYPE_DISCOVER)
+		reaction = _match_reaction(loaded_item, reaction)
+
+	if(reaction != FAIL)
```

Comment:
Any reason in particular why you're checking for `!= FAIL` specifically? If no, then we can just do `if(reaction)`

Reply:
Macro `FAIL` equals to `8` so this check is necessary.

Reply:
Gotcha